Neos.Neos/Classes/Controller/Service/NodesController.php

```diff
@@ -148,16 +151,27 @@ public function indexAction(
                 $entryNode = $subgraph->findNodeByAbsolutePath($nodePath);
             }
 
-            $nodes = !is_null($entryNode) ? $subgraph->findDescendantNodes(
-                $entryNode->aggregateId,
-                FindDescendantNodesFilter::create(
+            $nodes = Nodes::createEmpty();
+            if (!is_null($entryNode)) {
+                $filter = FindDescendantNodesFilter::create(
                     nodeTypes: NodeTypeCriteria::create(
                         NodeTypeNames::fromStringArray($nodeTypes),
                         NodeTypeNames::createEmpty()
                     ),
                     searchTerm: $searchTerm,
-                )
-            ) : [];
+                );
+                if (
+                    SearchTermMatcher::matchesNode($entryNode, $filter->searchTerm)
+                    && ExpandedNodeTypeCriteria::create($filter->nodeTypes, $contentRepository->getNodeTypeManager())
+                        ->matches($entryNode->nodeTypeName)
+                ) {
+                    // include the starting node if it matches
+                    $nodes = $nodes->append($entryNode);
+                }
+                $nodes = $nodes->merge(
+                    $subgraph->findDescendantNodes($entryNode->aggregateId, $filter)
+                );
```

Comment:
uh oh.. that adds so much "domain logic" to the controller.. Why did you decide to go for this approach instead of extending the filter and leaving implementation to the adapters?